### PR TITLE
i#5560 drmemtrace refactor: Split counting + i/o code

### PR DESCRIPTION
**Pull request by derekbruening**
Separates out two big pieces of tracer.cpp into their own files:
instruction counting into instr_counter.cpp and i/o into output.cpp.
Both still have some dependencies on some of the same TLS and globals
as the rest of the tracer, but shrinking the tracer.cpp file is still
of value, and the interfaces between the files are not huge.  Future
work could further modularize and isolate.

Adds a namespace, especially important now that we have global symbols
shared between files.

Issue: #5560, #4343

### CODE REVIEW

**Comment by derekbruening** [comment]
This shrinks tracer.cpp from 3405 lines down to 1881 lines.

The diff may not be easy to read: no changes are intended except for adding the namespaces; the rest is pure code movement.

**Comment by derekbruening** [comment]
Windows builds are failing frequently downloading doxygen (again)...

File: clients/drcachesim/tracer/tracer.h [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+/* XXX we don't have the inlining implemented yet for 32-bit architectures. */
+#endif
+
+#define INSTR_COUNT_LOCAL_UNIT 10000
+
+/***************************************************************************
+ * Buffer writing to disk.
+ */
+
+// NOCHECK move to output.h?
```
Comment (abhinav92003) [inline clients/drcachesim/tracer/tracer.h:175]: This was supposed to be addressed.

**Review by abhinav92003 — APPROVED** [review]
It's hard to do a more thorough review of the large diff. For more confidence, we should do more automated testing, like tracing on large apps.

File: clients/drcachesim/tracer/tracer.h (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/tracer.h:175]: Removed.  Ended up deciding to leave the content where it is.

**Comment by derekbruening** [comment]
32-bit windows failure is setthreadcontext #3213

**Comment by derekbruening** [comment]
I ran on some proprietary tests which are larger than the regression tests here including some larger "import tests" and the results match those before this PR.

**Pull request merged**